hypothesis-python/tests/cover/test_type_lookup.py

```diff
@@ -216,3 +219,58 @@ def test_uninspectable_from_type():
 def test_can_generate_from_all_registered_types(data, typ):
     value = data.draw(st.from_type(typ), label="value")
     assert isinstance(value, typ)
+
+
+T = TypeVar("T")
+
+
+class MyGeneric(Generic[T]):
+    def __init__(self, arg: T) -> None:
+        self.arg = arg
+
+
+def using_generic(instance: MyGeneric[T]) -> T:
+    return instance.arg
+
+
+def using_concrete_generic(instance: MyGeneric[int]) -> int:
+    return instance.arg
+
+
+def test_generic_origin_empty():
+    with pytest.raises(ResolutionFailed):
+        find_any(st.builds(using_generic))
+
+
+def test_generic_origin_with_type_args():
+    with pytest.raises(InvalidArgument):
+        st.register_type_strategy(MyGeneric[str], st.from_type(MyGeneric))
+
+
+def test_generic_origin_from_type():
+    try:
+        st.register_type_strategy(MyGeneric, st.builds(MyGeneric))
+        find_any(st.from_type(MyGeneric[T]))
+        find_any(st.from_type(MyGeneric[int]))
+        find_any(st.from_type(MyGeneric))
+    finally:
+        types._global_type_lookup.pop(MyGeneric)
+        st.from_type.__clear_cache()
```

Comment:
I think we're using this pattern often enough that we should define a context manager `registered_type(type_, resolve_to)` in `tests/common/utils.py` and `git grep _global_type_lookup.pop` for existing tests which should use it.

Reply:
What about general use? Sometimes we do recommend to use in-place in-function strategy registration.
It might be benefitial for end users as well:

```python
def my_test():
    with st.register_temp_type_strategy(...):
        do_my_stuff()
```

Reply:
I haven't seen any users asking for this yet, so I'd rather keep it as a private test helper for now.

We can easily add to the public API later, but removing anything takes a lot of work!

Reply:
New issue just for it: https://github.com/HypothesisWorks/hypothesis/issues/2544